soilmask.py

```diff
@@ -34,6 +50,151 @@ def gen_plant_mask(color_img: np.ndarray, kernel_size: int = 3) -> np.ndarray:
+    """
+    dimensions = np.shape(pixels)
+    if len(dimensions) == 2:
+        nrows, ncols = dimensions
+        channels = 1
+    else:
+        nrows, ncols, channels = dimensions
+
+    geotransform = (
+        gps_bounds[2], # upper-left x
+        (gps_bounds[3] - gps_bounds[2])/float(ncols), # W-E pixel resolution
+        0, # rotation (0 = North is up)
+        gps_bounds[1], # upper-left y
+        0, # rotation (0 = North is up)
+        -((gps_bounds[1] - gps_bounds[0])/float(nrows)) # N-S pixel resolution
+    )
+
+    # Create output GeoTIFF and set coordinates & projection
+    if asfloat:
+        dtype = gdal.GDT_Float32
+    else:
+        dtype = gdal.GDT_Byte
+
+    if compress:
+        output_raster = gdal.GetDriverByName('GTiff') \
+            .Create(str(out_path), ncols, nrows, channels, dtype, ['COMPRESS=LZW'])
+    else:
+        output_raster = gdal.GetDriverByName('GTiff') \
+            .Create(str(out_path), ncols, nrows, channels, dtype)
+
+    output_raster.SetGeoTransform(geotransform)
+    srs = osr.SpatialReference()
+    srs.ImportFromEPSG(4326) # google mercator
+    output_raster.SetProjection( srs.ExportToWkt() )
+
+    if channels == 3:
+        # typically 3 channels = RGB channels
+        # TODO: Something wonky w/ uint8s --> ending up w/ lots of gaps in data (white pixels)
+        output_raster.GetRasterBand(1).WriteArray(pixels[:,:,0].astype('uint8'))
+        output_raster.GetRasterBand(1).SetColorInterpretation(gdal.GCI_RedBand)
+        output_raster.GetRasterBand(1).FlushCache()
+        if nodata:
+            output_raster.GetRasterBand(1).SetNoDataValue(nodata)
+
+        output_raster.GetRasterBand(2).WriteArray(pixels[:,:,1].astype('uint8'))
+        output_raster.GetRasterBand(2).SetColorInterpretation(gdal.GCI_GreenBand)
+        output_raster.GetRasterBand(2).FlushCache()
+        if nodata:
+            output_raster.GetRasterBand(2).SetNoDataValue(nodata)
+
+        output_raster.GetRasterBand(3).WriteArray(pixels[:,:,2].astype('uint8'))
+        output_raster.GetRasterBand(3).SetColorInterpretation(gdal.GCI_BlueBand)
+        output_raster.GetRasterBand(3).FlushCache()
+        if nodata:
+            output_raster.GetRasterBand(3).SetNoDataValue(nodata)
+
+    elif channels > 1:
+        # TODO: Something wonky w/ uint8s --> ending up w/ lots of gaps in data (white pixels)
+        for chan in range(channels):
+            band = chan + 1
+            output_raster.GetRasterBand(band).WriteArray(pixels[:,:,chan].astype('uint8'))
+            output_raster.GetRasterBand(band).FlushCache()
+            if nodata:
+                output_raster.GetRasterBand(band).SetNoDataValue(nodata)
+    else:
+        # single channel image, e.g. temperature
+        output_raster.GetRasterBand(1).WriteArray(pixels)
+        output_raster.GetRasterBand(1).FlushCache()
+        if nodata:
+            output_raster.GetRasterBand(1).SetNoDataValue(nodata)
+
+    output_raster = None
```

Comment:
@Chris-Schnaufer This creates `output_raster` and then doesn't do anything with it. Is that right?

Reply:
The `output_raster` is used for its "side effects" of creating a georeferenced image on disk. Does that answer your question? This is direct from terrautils

Reply:
Sortof. Does that mean that if you don't set it to `None` that there will be memory leakage?

Reply:
Not at all https://python-textbok.readthedocs.io/en/1.0/Variables_and_Scope.html. That variable should go out of scope when the function returns and get cleaned up (at some point)

Reply:
OK. I know some Python libraries with C (or other native) bindings (e.g. [pandas](https://github.com/pandas-dev/pandas/issues?utf8=%E2%9C%93&q=is%3Aissue+memory+)) can have problems with dangling references.

I'm not familiar with this particular library, but I can't think of any other reason why someone would add a line like that unless they wanted to ensure that a leak doesn't happen.